### ISSUE TITLE
FR: Consider (re)mapping of subnets as an option

### ISSUE THREAD

**Issue by epposan** [tailscale/tailscale#13817]
### What are you trying to do?

There are plenty feature requests and reports that somehow are related to overlapping subnets and (remote) subnets getting priority over local net(work)s.

IMHO, if it would be possible to (re)map a subnet at the tailscale level, a lot of them could be worked-around more easily (and uniformly).

### How should we solve this?

With (re)mapping I mean, instead of advertising a **local** network (prefix+mask) as-is, advertise a new **virtual** network (with an identical mask) and **internally** translate it to and from the **actual** network's ip range.

If e.g. the **actual** network to be advertised is **192.168.0.0/24**, advertise it with an **offset**, e.g. as **192.168.10.0/24** and perform the translation **internally**

### What is the impact of not solving this?

More feature requests and issue reports related to overlapping subnets will be submitted ;-)

### Anything else?

(Re)mapping could be done

- at the **server** side (on the device advertising the subnet)
  in which case **all** clients are using the same (re)mapped subnet, or
- at the **client** side (on the device wanting to access the subnet),
  in which case **only** the client uses the (re)mapped subnet.

(Re)mapping can already be done today, however not (easily) on all platforms.

On many Linux based nodes it can be accomplished by using iptables' **NETMAP** option.

As an example, I use a (spare) Raspberry Pi, on which I

  - install a minimal O/S (in this case Debian)
  - install TailScale
  - enable SSH access
  
The remote location where I want to put this subnet router (for remote assistance) happens to have the **same** network prefix **192.168.0.0/24** as my local network, so, advertising it would bring me into trouble, as currently, unless action taken, the subnet would win.

To avoid this problem, I (re)map the remote network prefix **192.168.0.0/24** to something that does not conflict with my local network, e.g. **192.168.23.0/24**, and then advertise this (re)mapped **virtual** subnet

The following commands accomplish this:
```
    sudo iptables -t nat -A PREROUTING -i tailscale0 -d 192.168.23.0/24 -j NETMAP --to 192.168.0.0/24
    tailscale up --advertise-routes="192.168.23.0/24"
```
Looking at the routing table on my local (windows) machine after connecting to this tailnet, I get offered a route to **192.168.23.0/24**.

Running a ping to e.g. **192.168.23.1** on my local network, will now effectively run a ping to **192.168.0.1** on the **actual** remote network.

As stated above, unfortunately this cannot be easily and uniformly done on all platforms, so it would be a big plus if it could somehow be done **inside** tailscale.

I have no idea whether it can be implemented at all, or if there would be (negative) side-effects of this, but would appreciate to hear your thoughts on it.

Otherwise, a new **public** discussion to bring a resolution to the many subnet related feature request and issue reports, would be highly appreciated. (I'm pretty sure it's a topic frequently discussed **internally**)

Thanks for taking the time to read this ...

**Comment by agottardo** [comment]
> There are plenty feature requests and reports that somehow are related to overlapping subnets and (remote) subnets getting priority over local net(work)s.

Would you mind linking to such issues for context?

**Comment by epposan** [comment]
If you really insist ...

- 786      Force routing over Tailscale (prefer Tailscale routes over LAN)
- 826	    NAT (DNAT) conflicting subnets in advertise-routes
- 1227    When local route is available to a subnet, bypass tailscale subnet relay
- 1344    Unable to access hosts through tailscale on the same subnet as local machine
- 1413    Tailscale can't handle overlapping subnets
- 2590    Having multiple tailscale nodes in an advertised subnet may cause connectivity issues within that subnet
- 2697    Tailscale interface metric is so low Magic DNS always wins vs local network
- 2931    Allow mapping subnets or individual hosts on subnets to Tailscale addresses and MagicDNS names
- 6231    Tailsacle accepted subnet routes should not not have a higher priority than main table
- 7947    New option to accept routes at lower priority on Linux
- 9108    Tailscale Metric lower than LAN interface
- 12248  Inconsistent routing table metric settings on Windows
- 12657  Allow direct local network access from a Windows client with a subnet router on the same network
- 13817  Consider (re)mapping of subnets as an option

826      Even proposed a possible syntax in 2020: `--advertise-routes=192.168.23.0/24=192.68.0.0/24`

For the record, I have not put the `**#**' in front of the numbers to avoid the '**xxxx mentioned this**' entries, but if you want I can make it real **mentions**

**Comment by agottardo** [comment]
Oh, I totally wanted the mentions! It feeds into our automated tooling that sorts issues by priority, that's why I asked :) I added those for you.

**Comment by epposan** [comment]
Ah, I wasn't aware of that. Thanks.

Thinking about the syntax proposed in #826, another idea came to my mind.

Instead of using `=`, using symbols like `<=` and `=>` (or simply `<` and `>`) would make the mapping even more universal if the following already available options could be modified to accept them:

on the server side	`--advertise-routes=192.168.0.0/24=>192.168.23.0/24`
	
would advertise a **virtual** network 192.168.23.0/24 and internally map it to 192.168.0.0/24 on the server side

on the client side  `--accept-routes=192.168.23.0<=192.68.0.0/24`

would accept a (**real** or **virtual**) network 192.168.23.0 and internally map it to 192.168.0.0/24 on the client side

**Comment by AntoineDeveloper** [comment]
This would really be awesome

**Comment by gianfelicevincenzo** [comment]
At what point is the work? is something being done about it? before reading this post I had thought of creating a python script that would go to "turn off" and "turn on" the subnets of the various devices where I wanted the traffic to enter...

**Comment by iedame** [comment]
This would be amazing since not every network can be adjusted to avoid overlapping.

**Comment by AntoineDeveloper** [comment]
Honestly I really hope this issue will be resolved one day it would truely be amazing

**Comment by Ale1x** [comment]
Any update on this?

**Comment by mmogren** [comment]
$1000 bounty from me for this feature being added @agottardo

**Comment by Ale1x** [comment]
No updates?

**Comment by raffr85** [comment]
@mmogren Is your $1,000 bounty still available? My strongest programming experience is in Go, and I am interested in assessing this feature with disclosed Codex assistance.

Before starting implementation, could you clarify the acceptance and payment conditions: does the reward require an upstream merge, and would server-side IPv4 subnet remapping be the initial scope, or are client-side mapping and additional platforms required as well?

For maintainers: is an external contribution in this direction welcome, and is there an existing design or implementation effort I should coordinate with? I have not started a patch or committed to a delivery date.
